src/main/java/servlets/SearchServlet.java

```diff
@@ -0,0 +1,91 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+import com.google.appengine.api.blobstore.BlobKey;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.FetchOptions;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.Filter;
+import com.google.appengine.api.datastore.Query.FilterOperator;
+import com.google.appengine.api.datastore.Query.FilterPredicate;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import com.google.gson.Gson;
+import com.google.sps.data.Receipt;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.Spliterator;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/** Servlet that searches and returns matching receipts from datastore. */
+@WebServlet("/search-receipts")
+public class SearchServlet extends HttpServlet {
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String desiredLabel = request.getParameter("label");
+    ImmutableList<Receipt> receipts = getReceiptsWithMatchingLabel(desiredLabel);
+
+    Gson gson = new Gson();
+    response.setContentType("application/json;");
+    response.getWriter().println(gson.toJson(receipts));
+  }
+
+  /* Return ImmutableList of receipts from datastore with the same label as desiredLabel. */
+  private ImmutableList<Receipt> getReceiptsWithMatchingLabel(String desiredLabel) {
+    // Set filter to retrieve only receipts with label equal to desiredLabel.
+    Filter matchingLabels = new FilterPredicate("label", FilterOperator.EQUAL, desiredLabel);
+    Query query = new Query("Receipt");
+    query.setFilter(matchingLabels);
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+    // Convert matching receipt entities to receipt objects and add to return list.
+    return ImmutableList.copyOf(datastore.prepare(query)
+                                    .asList(FetchOptions.Builder.withDefaults())
+                                    .stream()
+                                    .map(entity -> { return createReceiptFromEntity(entity); })
```

Comment:
Normally, you should prefer method references to explicit lambdas, since the former usually result in shorter, cleaner code. For example:

`.map(this:: createReceiptFromEntity)`

Reply:
Thanks! I've never seen that notation before but that does make it cleaner